File: internal/webmail2/src/utils/smimeVault.ts
/**
 * Password-sealed S/MIME identity ("vault").
 *
 * The browser seals the private key and its certificate under a key derived from
 * the user's password (PBKDF2-SHA256, AES-256-GCM, both WebCrypto), and only the
 * sealed envelope goes to the server. The server stores it with the mailbox and
 * hands it back to its owner, so the identity is usable in every browser while the
 * server never holds anything it can open.
 */

/** VaultEnvelope is the sealed identity exactly as the server stores it. */
export interface VaultEnvelope {
  v: 1
  kdf: "PBKDF2-SHA256"
  iter: number
  salt: string
  iv: string
  ct: string
}

/** VaultContent is what the envelope seals: the key as PKCS#8 DER and the certificate. */
export interface VaultContent {
  keyPkcs8: string
  certPem: string
}

/** VAULT_ITERATIONS is the PBKDF2 work factor; the server refuses a lower one. */
export const VAULT_ITERATIONS = 600_000

// The format tag is authenticated with the ciphertext, so an envelope cannot be
// replayed as some other sealed format.
const AAD = new TextEncoder().encode("hermex-smime-vault-v1")

/** WrongVaultPassword is thrown when the envelope does not open under the password. */
export class WrongVaultPassword extends Error {
  constructor() {
    super("the password does not open the stored S/MIME key")
    this.name = "WrongVaultPassword"
  }
}

function toBase64(bytes: Uint8Array): string {
  let s = ""
  for (const b of bytes) s += String.fromCharCode(b)
  return btoa(s)
}

function fromBase64(s: string): Uint8Array<ArrayBuffer> {
  const bin = atob(s)
  const out = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i)
  return out
}

async function deriveKey(password: string, salt: Uint8Array<ArrayBuffer>, iter: number): Promise<CryptoKey> {
  const base = await crypto.subtle.importKey("raw", new TextEncoder().encode(password), "PBKDF2", false, ["deriveKey"])
  return crypto.subtle.deriveKey(
    { name: "PBKDF2", hash: "SHA-256", salt, iterations: iter },
    base,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  )
}

/** sealVault seals an identity under the password. */
export async function sealVault(content: VaultContent, password: string): Promise<VaultEnvelope> {
  const salt = crypto.getRandomValues(new Uint8Array(16))
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const key = await deriveKey(password, salt, VAULT_ITERATIONS)
  const plain = new TextEncoder().encode(JSON.stringify(content))
  const ct = new Uint8Array(await crypto.subtle.encrypt({ name: "AES-GCM", iv, additionalData: AAD }, key, plain))
  return { v: 1, kdf: "PBKDF2-SHA256", iter: VAULT_ITERATIONS, salt: toBase64(salt), iv: toBase64(iv), ct: toBase64(ct) }
}

/** openVault opens an envelope, throwing WrongVaultPassword when the password does not fit. */
export async function openVault(env: VaultEnvelope, password: string): Promise<VaultContent> {
  if (env.v !== 1 || env.kdf !== "PBKDF2-SHA256") throw new Error("unsupported S/MIME key format")
  const key = await deriveKey(password, fromBase64(env.salt), env.iter)
  let plain: ArrayBuffer
  try {
    plain = await crypto.subtle.decrypt({ name: "AES-GCM", iv: fromBase64(env.iv), additionalData: AAD }, key, fromBase64(env.ct))
  } catch {
    throw new WrongVaultPassword()
  }
  return JSON.parse(new TextDecoder().decode(plain)) as VaultContent
}

File: internal/webmail2api/smimevault.go
package webmail2api

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/http"

	"hermex/internal/logging"
)

// The S/MIME vault is a browser-mode identity the browser sealed under the user's
// password (PBKDF2-SHA256 into an AES-256-GCM key). The server stores the sealed
// envelope with the mailbox and returns it to its owner, so the identity works in
// every browser the owner signs in to, while the server never holds anything it
// can open.

// Vault limits. The work factor floor is the one the browser seals with, so a
// weaker envelope from a tampered client is refused; the ceiling keeps a stored
// envelope from making every later unlock hang. 64 KiB holds a PKCS#8 key and a
// certificate chain with room to spare.
const (
	vaultMinIterations = 600_000
	vaultMaxIterations = 10_000_000
	vaultMaxCiphertext = 64 << 10
)

// errBadVault reports an envelope that is not a sealed identity this server stores.
var errBadVault = errors.New("webmail2api: invalid S/MIME key vault")

// vaultEnvelope is the sealed identity as the browser writes it.
type vaultEnvelope struct {
	V    int    `json:"v"`
	KDF  string `json:"kdf"`
	Iter int    `json:"iter"`
	Salt string `json:"salt"`
	IV   string `json:"iv"`
	CT   string `json:"ct"`
}

// validVault checks an envelope's shape and returns it re-encoded, so what is
// stored is exactly the fields this format has. The ciphertext is not, and cannot
// be, checked beyond its size: only the owner's password opens it.
func validVault(raw []byte) ([]byte, error) {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.DisallowUnknownFields()
	var env vaultEnvelope
	if err := dec.Decode(&env); err != nil {
		return nil, errBadVault
	}
	if env.V != 1 || env.KDF != "PBKDF2-SHA256" || env.Iter < vaultMinIterations || env.Iter > vaultMaxIterations {
		return nil, errBadVault
	}
	if decodedLen(env.Salt) != 16 || decodedLen(env.IV) != 12 {
		return nil, errBadVault
	}
	if n := decodedLen(env.CT); n <= 0 || n > vaultMaxCiphertext {
		return nil, errBadVault
	}
	return json.Marshal(env)
}

// decodedLen is the length of a standard base64 value, -1 when it does not decode.
func decodedLen(s string) int {
	b, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		return -1
	}
	return len(b)
}

// handleGetSmimeVault returns the caller's sealed identity, 404 when they keep
// none on the server. The envelope is theirs alone and is never cached.
func (s *Server) handleGetSmimeVault(w http.ResponseWriter, r *http.Request) {
	st, c, ok := s.openStore(w, r)
	if !ok {
		return
	}
	defer st.Close()
	w.Header().Set("Cache-Control", "no-store")
	id, found, err := st.GetSmimeIdentity()
	if err != nil {
		logError("smime-vault", err, logging.Fields{"user": c.Email})
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "mailbox unavailable"})
		return
	}
	if !found || len(id.Vault) == 0 {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "no stored key"})
		return
	}
	writeJSON(w, http.StatusOK, json.RawMessage(id.Vault))
}
